c++/src/kj/compat/http.c++

```diff
@@ -3714,13 +4076,96 @@ public:
     });
   }
 
-private:
-  HttpInputStreamImpl httpInput;
-  HttpOutputStream httpOutput;
-  kj::Own<AsyncIoStream> ownStream;
-  HttpClientSettings settings;
-  kj::Maybe<kj::Promise<void>> closeWatcherTask;
-  bool upgraded = false;
+  kj::Promise<ConnectResponse> connect(kj::StringPtr host, const HttpHeaders& headers) override {
+    KJ_REQUIRE(!upgraded,
+        "can't make further requests on this HttpClient because it has been or is in the process "
+        "of being upgraded");
+    KJ_REQUIRE(!closed,
+        "this HttpClient's connection has been closed by the server or due to an error");
+    KJ_REQUIRE(httpOutput.canReuse(),
+        "can't start new request until previous request body has been fully written");
+
+    closeWatcherTask = nullptr;
+
+    kj::StringPtr connectionHeaders[HttpHeaders::CONNECTION_HEADERS_COUNT];
+    kj::String lengthStr;
+
+    // TODO(now): Make sure that forbidden headers are not included...
+
+    httpOutput.writeHeaders(headers.serializeConnect(host, connectionHeaders));
+
+    auto id = ++counter;
+
+    return httpInput.readResponseHeaders().then(
+        [this,id](HttpHeaders::ResponseOrProtocolError&& responseOrProtocolError)
+            -> ConnectResponse {
+      KJ_SWITCH_ONEOF(responseOrProtocolError) {
+        KJ_CASE_ONEOF(response, HttpHeaders::Response) {
+          auto& responseHeaders = httpInput.getHeaders();
+          if (response.statusCode < 200 || response.statusCode >= 300) {
+            // Any statusCode that is not in the 2xx range in interpreted
+            // as an HTTP response. Any status code in the 2xx range is
+            // interpreted as a successful CONNECT response.
+            closed = true;
+            kj::Own<kj::AsyncInputStream> body = heap<NullInputStream>();
+            return ConnectResponse {
+              .statusCode = response.statusCode,
+              .statusText = response.statusText,
+              .headers = &responseHeaders,
+              .connectionOrBody = kj::mv(body),
+            };
+          }
+
+          auto& stream = *ownStream;
+          kj::Own<kj::AsyncIoStream> io = heap<ConnectPipeEnd>(
+              kj::Own<kj::AsyncIoStream>(&stream, kj::NullDisposer::instance),
```

Comment:
I don't think you can just use the underlying `stream` here because you'll miss any bytes that were left over in the header buffer.